🍾  Add env based toleration to scheduled pods.

prasrvenkat opened the pull request:
## What

Ability to add tolerations optionally so we can restrict where we schedule pods to fix https://github.com/airbytehq/airbyte/issues/5536

Env config
```
WORKER_POD_TOLERATIONS="key=airbyte,operator=Equal,value=false,effect=NoSchedule"
```

... adds the toleration to scheduled PODs (the first one)

![image](https://user-images.githubusercontent.com/47996246/130522073-f2574979-6336-472b-9433-a4e644649e00.png)

## How
* Add env variable `WORKER_POD_TOLERATIONS` with a value format that is easy to parse and use.
* Parse that value optionally and provide as input to PodBuilder.
* I originally started adding it similar to `ResourceRequirements` and ended up modifying more places and finally `StandardSyncInput` model as well. I believe that level of tuning is not needed for this specifically, let me know if that is what we prefer, happy to revert to it.

## Recommended reading order
1. `WorkerPodToleration.java`
2. `EnvConfigs.java#getWorkerPodTolerations`
3. `EnvConfigsTest.java#testWorkerPodTolerations`
4. `KubePodProcess.java#buildPodTolerations`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>


<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

davinchia commented on airbyte-config/models/src/test/java/io/airbyte/config/EnvConfigsTest.java:
```diff
@@ -179,4 +180,27 @@ void testTrackingStrategy() {
   }
 
+  @Test
+  void testWorkerPodTolerations() {
```
Nice!

davinchia commented on airbyte-config/models/src/main/java/io/airbyte/config/WorkerPodToleration.java:
```diff
@@ -0,0 +1,54 @@
+
+import java.util.Objects;
+
+// represents a minimal io.fabric8.kubernetes.api.model.Toleration
```
Can this be a proper doc string?

prasrvenkat replied:
Yeah sure

davinchia commented on airbyte-config/models/src/main/java/io/airbyte/config/EnvConfigs.java:
```diff
@@ -232,6 +238,38 @@ public WorkspaceRetentionConfig getWorkspaceRetentionConfig() {
 
+  @Override
+  public List<WorkerPodToleration> getWorkerPodTolerations() {
+    // Format:
```
Nice comment.

Can this be a proper function doc string?

prasrvenkat replied:
Yeah sure

davinchia commented on airbyte-config/models/src/main/java/io/airbyte/config/EnvConfigs.java:
```diff
@@ -232,6 +238,38 @@ public WorkspaceRetentionConfig getWorkspaceRetentionConfig() {
+  @Override
+  public List<WorkerPodToleration> getWorkerPodTolerations() {
+    // Format:
+    // Tolerations separated by ;;
```
What's the reason for using `;;` as the delimiter?

prasrvenkat replied:
No specific reason, I usually use it in similar parsing. I can change to just one `;` if you prefer

prasrvenkat replied:
I changed to just ; btw

davinchia commented on airbyte-workers/src/main/java/io/airbyte/workers/process/KubePodProcess.java:
```diff
@@ -243,6 +246,21 @@ private static void waitForInitPodToRun(KubernetesClient client, Pod podDefiniti
   }
 
+  private Toleration[] buildPodTolerations(List<WorkerPodToleration> tolerations) {
+    if (tolerations == null || tolerations.isEmpty()) {
+      return null;
+    } else {
+      return tolerations.stream().map(workerPodToleration ->
+              new TolerationBuilder()
+                  .withKey(workerPodToleration.getKey())
+                  .withEffect(workerPodToleration.getEffect())
+                  .withOperator(workerPodToleration.getOperator())
+                  .withValue(workerPodToleration.getValue())
+                  .build())
+          .toArray(Toleration[]::new);
+    }
+  }
+
```
```suggestion
    if (tolerations == null || tolerations.isEmpty()) {
      return null;
    }
    return tolerations.stream().map(workerPodToleration ->
            new TolerationBuilder()
                .withKey(workerPodToleration.getKey())
                .withEffect(workerPodToleration.getEffect())
                .withOperator(workerPodToleration.getOperator())
                .withValue(workerPodToleration.getValue())
                .build())
        .toArray(Toleration[]::new);
    }
  }

```

davinchia replied:
Since we return early, we can get rid of the else statement for readability.

davinchia commented on airbyte-config/models/src/main/java/io/airbyte/config/EnvConfigs.java:
```diff
@@ -232,6 +238,38 @@ public WorkspaceRetentionConfig getWorkspaceRetentionConfig() {
+    // we need at-least key, effect and operator to be present. value is optional
+    // we do not
+    String tolerations = getEnvOrDefault(WORKER_POD_TOLERATIONS, "");
+    return Strings.isNullOrEmpty(tolerations) ? List.of() : Splitter.on(";;")
+        .splitToStream(tolerations)
+        .filter(tolerationStr -> !Strings.isNullOrEmpty(tolerationStr))
+        .map(tolerationStr -> {
+          Map<String, String> tolerationMap = Splitter.on(",")
+              .splitToStream(tolerationStr)
+              .map(s -> s.split("="))
+              .collect(Collectors.toMap(s -> s[0], s -> s[1]));
+          if (tolerationMap.containsKey("key") && tolerationMap.containsKey("effect") && tolerationMap.containsKey("operator")) {
+            return new WorkerPodToleration(tolerationMap.get("key"),
+                tolerationMap.get("effect"),
+                tolerationMap.get("value"),
+                tolerationMap.get("operator"));
+          } else {
+            LOGGER.warn("Ignoring toleration {}, missing one of key,effect or operator",
+                tolerationStr);
+            return null;
+          }
+        })
+        .filter(Objects::nonNull)
+        .collect(Collectors.toList());
```
This block is tough to read.

What about splitting into different statements? One natural split is 1) parse tolerations from env var 2) map tolerations into `WorkerPodToleration` objects.

The map function can also be pulled out into a separate function to keep things simple. What do you think?

prasrvenkat replied:
Yeah sounds good, will do.

davinchia reviewed (commented):
Thanks for the contribution @prasrvenkat ! So great!

Some minor comments on readability. Let me know what you think.

From the screenshot, looks like you tested this locally?

prasrvenkat commented:
> Thanks for the contribution @prasrvenkat ! So great!
> 
> Some minor comments on readability. Let me know what you think.
> 
> From the screenshot, looks like you tested this locally?

Addressing them right now, yes I tested in one of our GKE clusters

prasrvenkat commented:
> > Thanks for the contribution @prasrvenkat ! So great!
> > Some minor comments on readability. Let me know what you think.
> > From the screenshot, looks like you tested this locally?
> 
> Addressing them right now, yes I tested in one of our GKE clusters

@davinchia Addressed your comments, thanks for review. Let me know if that looks good.

davinchia commented:
@prasrvenkat looks great! Thanks for the quick turn around! Will merge in today!
